app/src/main/java/com/meewii/rateconverter/business/ExchangeRateRepository.kt

```diff
@@ -89,37 +71,31 @@ class ExchangeRateRepository @Inject constructor(
   }
 
   @VisibleForTesting
-  internal fun getApiRates(base: String): Single<RateList> {
+  internal fun getApiRates(base: String): Flowable<RateList> {
     return service.getRatesForBase(base)
       .subscribeOn(Schedulers.io())
-      .map { response ->
-        if (response.errorMessage != null) RateList.Error(null, response.errorMessage)
-        else if (response.base == null || response.rates == null || response.date == null) {
-          RateList.Error(InvalidResponseException("The Response is invalid, missing rates, date or base currency"))
+      .toFlowable()
+      .flatMap { response ->
+        if (response.errorMessage != null) {
+          Flowable.error(ResponseErrorException(response.errorMessage))
+        } else if (response.base == null || response.rates == null || response.date == null) {
+          Flowable.error(InvalidResponseException("The Response is invalid, missing rates, date or base currency"))
         } else {
           val filteredRates = response.rates.toMutableMap().apply { remove(base) }
           // The European Central Bank usually rates only once a day around 16:00 CET
           val ldt = LocalDateTime.of(response.date, LocalTime.of(16, 0))
           persistExchangeRates(response.base, filteredRates, ldt)
-          RateList.Success(toCurrencyList(filteredRates))
+          Flowable.just(RateList.Success(response.base, toCurrencyList(filteredRates)))
```

Comment:
Here is another suggestion: 
According to clean code principles, the repository should not modify data model that passes to the UI. Currently as far as I understand, if the response is successful and we can show the response in the UI, we are flatmapping a Success view state. 
I would suggest the following => For me it makes sense to have a testable, injectable mapper class, that transforms RateList to RateListViewState, which has a success, loading and error states. This way if the response changes, I can just adjust my mapper class, adjust my tests accordingly so that I will never forget anything and go on with my life because the view will always act the same :D 
What do you think?

Reply:
Because of the additional sorting logic in The ViewModel, I prefer to keep the state of the view separated from the State of the response. I simplify the response though to get rid of a state of RateList completely.